npm/pod.go

```
@@ -0,0 +1,140 @@
+	log.Printf("POD CREATED: %s/%s/%s%+v%s\n", podNs, podName, podNodeName, podLabels, podIP)
+
+	// Add the pod to ipset
+	ipsMgr := npMgr.nsMap[util.KubeAllNamespacesFlag].ipsMgr
+	// Add the pod to its namespace's ipset.
+	log.Printf("Adding pod %s to ipset %s\n", podIP, podNs)
+	if err := ipsMgr.AddToSet(podNs, podIP); err != nil {
+		log.Printf("Error adding pod to namespace ipset.\n")
+	}
+
+	// Add the pod to its label's ipset.
+	var labelKeys []string
+	for podLabelKey, podLabelVal := range podLabels {
+		//Ignore pod-template-hash label.
+		if strings.Contains(podLabelKey, util.KubePodTemplateHashFlag) {
+			continue
+		}
+
+		labelKey := util.KubeAllNamespacesFlag + "-" + podLabelKey + ":" + podLabelVal
+		log.Printf("Adding pod %s to ipset %s\n", podIP, labelKey)
+		if err := ipsMgr.AddToSet(labelKey, podIP); err != nil {
+			log.Printf("Error adding pod to label ipset.\n")
+			return err
+		}
+		labelKeys = append(labelKeys, labelKey)
+	}
+
+	ns, err := newNs(podNs)
+	if err != nil {
+		log.Printf("Error creating namespace %s\n", podNs)
+	}
+	npMgr.nsMap[podNs] = ns
+
+	return nil
+}
+
+// UpdatePod handles update pod.
+func (npMgr *NetworkPolicyManager) UpdatePod(oldPodObj, newPodObj *corev1.Pod) error {
+	if !isValidPod(newPodObj) {
+		return nil
+	}
+
+	oldPodObjNs := oldPodObj.ObjectMeta.Namespace
+	oldPodObjName := oldPodObj.ObjectMeta.Name
+	oldPodObjPhase := oldPodObj.Status.Phase
+	oldPodObjIP := oldPodObj.Status.PodIP
+	newPodObjNs := newPodObj.ObjectMeta.Namespace
+	newPodObjName := newPodObj.ObjectMeta.Name
+	newPodObjPhase := newPodObj.Status.Phase
+	newPodObjIP := newPodObj.Status.PodIP
+
+	log.Printf(
+		"POD UPDATED. %s %s %s %s %s %s %s %s\n",
+		oldPodObjNs, oldPodObjName, oldPodObjPhase, oldPodObjIP, newPodObjNs, newPodObjName, newPodObjPhase, newPodObjIP,
+	)
+
+	npMgr.DeletePod(oldPodObj)
+
+	if newPodObj.ObjectMeta.DeletionTimestamp == nil && newPodObj.ObjectMeta.DeletionGracePeriodSeconds == nil {
+		npMgr.AddPod(newPodObj)
```

If control doesn't go to line 96, we won't create a new pod. Should we return an error in that case?